Update master service ports and type via controller.

abutcher opened the pull request:
Since master service ports and type are configurable it makes sense to ensure that these attributes are correctly set in the controller loop.

/cc @smarterclayton @liggitt

abutcher committed:
Update master service ports and type via controller.

liggitt commented on pkg/master/controller.go:
```diff
@@ -117,6 +117,9 @@ func (c *Controller) UpdateKubernetesService() error {
 		if err := c.CreateMasterServiceIfNeeded("kubernetes", c.ServiceIP, servicePorts, serviceType); err != nil {
 			return err
 		}
+		if err := c.SetService("kubernetes", servicePorts, serviceType); err != nil {
```
seems like we'd only want a single function `CreateOrUpdateMasterServiceIfNeeded`

liggitt replied:
how will this work if multiple masters have differing "extra ports"? how would you go about updating that setting if you needed to? You couldn't restart all the masters simultaneously, and you wouldn't want outdated masters stomping the new setting

abutcher replied:
All masters must use the same ports and admins injecting extra ports would need to be aware of this. I've self-doc'd that above the `SetService` function in case someone runs into it - similarly, extra endpoints must match on all api servers as well.

abutcher replied:
I can move the `SetService` call into what would be `CreateOrUpdateMasterServiceIfNeeded` but I'd like to stick with the `SetService` function as a compartmental thing to test.

abutcher replied:
Maybe a different name and move the logic into `CreateOrUpdate...` ? Something like, if the service already exists make sure it checks out and update it if necessary.

abutcher commented:
@liggitt To avoid stomping the existing ports on either the master service or endpoints, missing ports are retained from the format checks and appended if any were missing. This saves us from having to dedupe the port list prior to updating.

abutcher commented:
@smarterclayton Updated

smarterclayton commented:
LGTM unless @liggitt has other feedback

liggitt commented on pkg/master/controller.go:
```diff
@@ -296,13 +302,67 @@ func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.Endpoint
+	formatCorrect = true
+	for _, port := range ports {
+		contains := false
+		for _, servicePort := range s.Spec.Ports {
```
If this is a partial match (port number matches but name doesn't, or vice-versa), `missingPorts` will be returned which overlap with existing ports. When you try to update the service, validation will fail on duplicate portNames or duplicate ports. Be sure to add tests for this case.

Limiting the mutations to setting the type and adding missing ports is better, but I still don't see how this prevents thrash between masters if a port name or number changes (e.g. we add the "dns" name to a port and an old master keeps trying to update the service to remove that name).

abutcher replied:
I think that `missingPorts` should take precedence in a collision because we expected them to be there but trash may be unavoidable.

liggitt replied:
so in that case, just appending missing ports to the existing service ports could cause problems... you need to update the service in context, replacing conflicting ports and appending missing ports

abutcher commented:
@liggitt Updated to account for collisions where `missingPorts` take precedence. Added some tests to show that we take the `missingPort` in a collision.

abutcher commented:
@liggitt @smarterclayton we're going to need this for HA. Can either of you take another look?

smarterclayton commented:
I have no more comments, Jordan has sign off

liggitt commented on pkg/master/controller.go:
```diff
@@ -179,11 +179,15 @@ func createEndpointPortSpec(endpointPort int, endpointPortName string, extraEndp
 		// The service already exists.
-		return nil
+		if svc, updated := updateMasterServiceIfNeeded(s, servicePorts, serviceType); updated {
+			glog.Warningf("Resetting spec for master service %q to %v", serviceName, svc)
```
`glog.Infof("Updating master service %q to %#v", ...`

liggitt replied:
mm, I see the endpoints were already `Warningf`... don't feel strongly if we want to keep that

liggitt commented on pkg/master/controller.go:
```diff
@@ -179,11 +179,15 @@ func createEndpointPortSpec(endpointPort int, endpointPortName string, extraEndp
+	if s, err := c.ServiceRegistry.GetService(ctx, serviceName); err == nil {
 		// The service already exists.
-		return nil
+		if svc, updated := updateMasterServiceIfNeeded(s, servicePorts, serviceType); updated {
```
rename? this doesn't actually do the update, just computes it:
`if svc, update := getMasterServiceUpdateIfNeeded(...); update { ...`

liggitt commented on pkg/master/controller_test.go:
```diff
@@ -365,3 +366,288 @@ func TestSetEndpoints(t *testing.T) {
+			},
+		},
+		{
+			testName:    "service definition incorrect port",
```
better name, hard to determine intent, maybe "extra port and missing port"

abutcher replied:
I left this name since it makes sense w/o missing ports.

liggitt commented on pkg/master/controller_test.go:
```diff
@@ -365,3 +366,288 @@ func TestSetEndpoints(t *testing.T) {
+				},
+			},
+			expectUpdate: nil,
+		},
```
add a test case to make sure an extra port is tolerated when everything else matches

liggitt commented on pkg/master/controller.go:
```diff
@@ -296,13 +303,104 @@ func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.Endpoint
+	for _, servicePort := range servicePorts {
+		collision := false
+		for _, missingPort := range missingPorts {
+			if (missingPort.Name == servicePort.Name) || (missingPort.Port == servicePort.Port && missingPort.Protocol == servicePort.Protocol) || (missingPort.TargetPort == servicePort.TargetPort && missingPort.Protocol == servicePort.Protocol) {
```
is name optional? if both are empty, that's not a collision, right?

liggitt replied:
looks like name is optional if there's a single port? that means if this starts with a single nameless non-conflicting port, and appends other named non-conflicting ports, the service will fail validation because the original port had no name.

liggitt commented on pkg/master/controller.go:
```diff
@@ -296,13 +303,104 @@ func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.Endpoint
+// filterServicePorts combines missing service ports with existing service ports.
+// Missing ports take precedence over existing ports when a collision occurs.
+func filterServicePorts(servicePorts []api.ServicePort, missingPorts []api.ServicePort) []api.ServicePort {
+	newServicePorts := servicePorts[:0]
```
it would be easier to read to start with a new slice, and append the originals if non-conflicting

abutcher commented:
@liggitt Changes are in for only reconciling service / endpoint definitions on start. I removed a lot of the port maintenance cruft since we expect ports to match exactly on start. Also added a test w/o reconcile to show that the master service would not be updated unless reconcile is set to true.

liggitt commented on pkg/master/controller.go:
```diff
@@ -179,11 +185,15 @@ func createEndpointPortSpec(endpointPort int, endpointPortName string, extraEndp
-	if _, err := c.ServiceRegistry.GetService(ctx, serviceName); err == nil {
-		// The service already exists.
-		return nil
+	if s, err := c.ServiceRegistry.GetService(ctx, serviceName); err == nil && reconcile {
```
needs to be:
```
if s, err := c.ServiceRegistry.GetService(ctx, serviceName); err == nil {
  // The service already exists.
  if reconcile {
    ...
  }
  return nil
}
```
also, add tests that catch this case

liggitt commented on pkg/master/controller.go:
```diff
@@ -114,12 +118,14 @@ func (c *Controller) UpdateKubernetesService() error {
-			return err
+		if reconcile {
+			endpointPorts := createEndpointPortSpec(c.PublicServicePort, "https", c.ExtraEndpointPorts)
+			if err := c.SetEndpoints("kubernetes", c.PublicIP, endpointPorts); err != nil {
```
doesn't SetEndpoints still need to be called in the loop in case the endpoint gets deleted? probably pass `reconcile` to it so it (and maybe rename to `reconcilePorts`)

liggitt replied:
I was hoping this would look more like a couple new blocks of very strict checks that were done when `reconcilePorts` was true, and the rest of the code would stay as-is

liggitt commented on pkg/master/controller.go:
```diff
@@ -237,23 +247,21 @@ func (c *Controller) SetEndpoints(serviceName string, ip net.IP, endpointPorts [
 	// First, determine if the endpoint is in the format we expect (one
-	// subset, ports matching endpointPorts, N IP addresses).
+	// subset, endpointPorts are present, N IP addresses).
 	formatCorrect, ipCorrect := checkEndpointSubsetFormat(e, ip.String(), endpointPorts, c.MasterCount)
```
what about 
```
formatCorrect, ipCorrect, portsCorrect := checkEndpointSubsetFormat(e, ip.String(), endpointPorts, c.MasterCount, reconcilePorts)
if !formatCorrect {
  // existing egregious code
  return update
} else if !ipCorrect || !portsCorrect {
  if !ipCorrect {
    // append IP, repack
  }
  if !portsCorrect {
    // reset ports
  }
  return update
}
return nil
```

abutcher commented:
@liggitt Comments addressed. I plumbed reconcile into the endpoints check but not the service check. If the service type is incorrect then the ports are incorrect (from the `nodePort` setting) and we'll need to reconcile either way. Added a test for create when the master service doesn't exist and added one for an endpoint update without `reconcilePorts`.

liggitt commented on pkg/master/controller.go:
```diff
@@ -271,38 +286,80 @@ func (c *Controller) SetEndpoints(serviceName string, ip net.IP, endpointPorts [
+func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.EndpointPort, count int, reconcilePorts bool) (formatCorrect bool, ipCorrect bool, portsCorrect bool) {
 	if len(e.Subsets) != 1 {
-		return false, false
+		return false, true, true
```
`false,false,false`?

liggitt commented on pkg/master/controller.go:
```diff
@@ -271,38 +286,80 @@ func (c *Controller) SetEndpoints(serviceName string, ip net.IP, endpointPorts [
-			return false, false
+		for i, port := range ports {
+			if port != sub.Ports[i] {
+				return true, false, false
```
returning false for `ipCorrect` here could make us append our IP when it's not actually needed... you need to correctly calculate both `portsCorrect` and `ipCorrect`

```
portsCorrect = true
if reconcilePorts {
  for i, port := range ports {
    if len(sub.Ports) <= i || port != sub.Ports[i] {
      portsCorrect = false
      break
    }
  }
}

// calculate ipCorrect

return true, ipCorrect, portsCorrect
```

liggitt replied:
add a test case to make sure incorrect ports, but a correct IP, doesn't result in a double-appended IP

abutcher commented:
@liggitt Updated. There are already several tests with correct IP and incorrect ports, like "existing endpoints wrong port".

liggitt commented:
>  There are already several tests with correct IP and incorrect ports, like "existing endpoints wrong port".

With reconcile? The previous code would have returned ipCorrect=false, which a test should have caught a double-append for, right?

abutcher commented:
@liggitt Yep, with reconcile. One is added but ips after ours are removed so that the list eventually becomes correct I think.

liggitt commented on pkg/master/controller_test.go:
```diff
@@ -305,15 +306,53 @@ func TestSetEndpoints(t *testing.T) {
+		endpoints         *api.EndpointsList
+		expectUpdate      *api.Endpoints // nil means none expected
+	}{
 		{
```
at least one non-reconcile test with an incorrect IP, and one with an incorrect format to demonstrate non-port-reconcile case still updates

liggitt commented:
test comment, then LGTM

abutcher commented:
@liggitt Added two non-reconcile endpoint tests: one with wrong ports/wrong ip in which we only expect the ip to be updated and another test where the format is wrong but the subsets are corrected.

abutcher commented:
@liggitt and this still needs a test tag

liggitt commented:
@k8s-bot test this

liggitt commented:
@k8s-bot unit test this

liggitt commented on pkg/master/controller.go:
```diff
@@ -271,38 +286,80 @@ func (c *Controller) SetEndpoints(serviceName string, ip net.IP, endpointPorts [
+	portsCorrect = true
+	if reconcilePorts {
+		for i, port := range ports {
+			if len(sub.Ports) <= i || port != sub.Ports[i] {
```
my mistake, you need to do the length check separately, or an empty `ports` slice will never reach this check

abutcher commented:
@liggitt Fixed.

liggitt commented:
LGTM

wojtek-t commented:
@k8s-bot unit test this please
